In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [6]:
# Get SEABORN datasets
# Use get_dataset_names() to see a list of available datasets

sns.get_dataset_names()

['anagrams',
 'anscombe',
 'attention',
 'brain_networks',
 'car_crashes',
 'diamonds',
 'dots',
 'exercise',
 'flights',
 'fmri',
 'gammas',
 'geyser',
 'iris',
 'mpg',
 'penguins',
 'planets',
 'taxis',
 'tips',
 'titanic']

In [7]:
# Load Titanic Dataset

ttn=sns.load_dataset('titanic')
ttn

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [8]:
# Get quick view of data
ttn.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [9]:
ttn.tail()

survived  pclass     sex   age  sibsp  parch   fare embarked   class  \
886         0       2    male  27.0      0      0  13.00        S  Second   
887         1       1  female  19.0      0      0  30.00        S   First   
888         0       3  female   NaN      1      2  23.45        S   Third   
889         1       1    male  26.0      0      0  30.00        C   First   
890         0       3    male  32.0      0      0   7.75        Q   Third   

       who  adult_male deck  embark_town alive  alone  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True

In [10]:
ttn.sample(10)

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
14          0       3  female  14.0      0      0   7.8542        S   Third   
490         0       3    male   NaN      1      0  19.9667        S   Third   
384         0       3    male   NaN      0      0   7.8958        S   Third   
535         1       2  female   7.0      0      2  26.2500        S  Second   
885         0       3  female  39.0      0      5  29.1250        Q   Third   
665         0       2    male  32.0      2      0  73.5000        S  Second   
754         1       2  female  48.0      1      2  65.0000        S  Second   
107         1       3    male   NaN      0      0   7.7750        S   Third   
491         0       3    male  21.0      0      0   7.2500        S   Third   
376         1       3  female  22.0      0      0   7.2500        S   Third   

       who  adult_male deck  embark_town alive  alone  
14   child       False  NaN  Southampton    no   True  
490    man        True  NaN  Southampton    no  False  
384    man        True  NaN  Southampton    no   True  
535  child       False  NaN  Southampton   yes  False  
885  woman       False  NaN   Queenstown    no  False  
665    man        True  NaN  Southampton    no  False  
754  woman       False  NaN  Southampton   yes  False  
107    man        True  NaN  Southampton   yes   True  
491    man        True  NaN  Southampton    no   True  
376  woman       False  NaN  Southampton   yes   True

# Understand columns and data
# https://www.kaggle.com/c/titanic/data
Variable	Definition	Key
survival	Survival	0 = No, 1 = Yes
pclass	Ticket class	1 = 1st, 2 = 2nd, 3 = 3rd
sex	Sex	
Age	Age in years	
sibsp	# of siblings / spouses aboard the Titanic	
parch	# of parents / children aboard the Titanic	
ticket	Ticket number	
fare	Passenger fare	
cabin	Cabin number	
embarked	Port of Embarkation

In [11]:
# Get name of the columns

ttn.columns

# This allows to check if column names have any spaces which need to be trimmed. Info cannot make out.

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [12]:
# Get  number of rows, columns
ttn.shape

(891, 15)

In [13]:
# Get number of elements = rows mutliplied by columns (=891x15)

ttn.size

13365

In [14]:
# Get summary data

ttn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [15]:
# Get summary statistics of all numerical columns
ttn.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [16]:
# Get the number of Null values(NaN)
ttn.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [97]:
# Get the number of Null values(NaN) (alternate method using .isnull())
ttn.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

a) The ages of 177 passengers were unknown
b) Even though ages were not known yet classification of child, woman and man(who column) and adult_male,true or false exists.
c) Only two records dont have embarkation town.

# filter dataframe = df[condition]
### remember square brackets

In [69]:
# Filtering the data frame for nulls in embark town column.
ttn[ttn['embark_town'].isna()]

survived  pclass     sex   age  sibsp  parch  fare embarked  class  \
61          1       1  female  38.0      0      0  80.0      NaN  First   
829         1       1  female  62.0      0      0  80.0      NaN  First   

       who  adult_male deck embark_town alive  alone  
61   woman       False    B         NaN   yes   True  
829  woman       False    B         NaN   yes   True

## df multiple conditions each in round bracket

In [81]:
# Filtering data frame for age greater than 60 and females only. (Using AND operator)
ttn[(ttn['age'] > 60) & (ttn['sex'] == 'female') ]

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
275         1       1  female  63.0      1      0  77.9583        S  First   
483         1       3  female  63.0      0      0   9.5875        S  Third   
829         1       1  female  62.0      0      0  80.0000      NaN  First   

       who  adult_male deck  embark_town alive  alone  
275  woman       False    D  Southampton   yes  False  
483  woman       False  NaN  Southampton   yes   True  
829  woman       False    B          NaN   yes   True

In [82]:
# Filtering data frame for age greater than 60 or females only. (Using OR operator)
ttn[(ttn['age'] > 60) | (ttn['sex'] == 'female') ]

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
8           1       3  female  27.0      0      2  11.1333        S   Third   
9           1       2  female  14.0      1      0  30.0708        C  Second   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
880         1       2  female  25.0      0      1  26.0000        S  Second   
882         0       3  female  22.0      0      0  10.5167        S   Third   
885         0       3  female  39.0      0      5  29.1250        Q   Third   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   

       who  adult_male deck  embark_town alive  alone  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
8    woman       False  NaN  Southampton   yes  False  
9    child       False  NaN    Cherbourg   yes  False  
..     ...         ...  ...          ...   ...    ...  
880  woman       False  NaN  Southampton   yes  False  
882  woman       False  NaN  Southampton    no   True  
885  woman       False  NaN   Queenstown    no  False  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  

[333 rows x 15 columns]

In [17]:
# Number or count of unique values in the column survived
ttn['survived'].nunique()

2

In [18]:
# Show unique values in the column survived
ttn['survived'].unique()

array([0, 1], dtype=int64)

In [20]:
ttn['embark_town'].unique()

array(['Southampton', 'Cherbourg', 'Queenstown', nan], dtype=object)

In [21]:
# Number or count of unique values in the Data frame
ttn.nunique()

survived         2
pclass           3
sex              2
age             88
sibsp            7
parch            7
fare           248
embarked         3
class            3
who              3
adult_male       2
deck             7
embark_town      3
alive            2
alone            2
dtype: int64

a) There were only 891 passenger but there were 248 Fare types despite having only three classes.

In [28]:
# Find unique values and sort the list
farelist = ttn['fare'].unique()
farelist.sort()
farelist

array([  0.    ,   4.0125,   5.    ,   6.2375,   6.4375,   6.45  ,
         6.4958,   6.75  ,   6.8583,   6.95  ,   6.975 ,   7.0458,
         7.05  ,   7.0542,   7.125 ,   7.1417,   7.225 ,   7.2292,
         7.25  ,   7.3125,   7.4958,   7.5208,   7.55  ,   7.6292,
         7.65  ,   7.725 ,   7.7292,   7.7333,   7.7375,   7.7417,
         7.75  ,   7.775 ,   7.7875,   7.7958,   7.8   ,   7.8292,
         7.8542,   7.875 ,   7.8792,   7.8875,   7.8958,   7.925 ,
         8.0292,   8.05  ,   8.1125,   8.1375,   8.1583,   8.3   ,
         8.3625,   8.4042,   8.4333,   8.4583,   8.5167,   8.6542,
         8.6625,   8.6833,   8.7125,   8.85  ,   9.    ,   9.2167,
         9.225 ,   9.35  ,   9.475 ,   9.4833,   9.5   ,   9.5875,
         9.825 ,   9.8375,   9.8417,   9.8458,  10.1708,  10.4625,
        10.5   ,  10.5167,  11.1333,  11.2417,  11.5   ,  12.    ,
        12.275 ,  12.2875,  12.35  ,  12.475 ,  12.525 ,  12.65  ,
        12.875 ,  13.    ,  13.4167,  13.5   ,  13.7917,  13.8

In [30]:
# List of unique values and the count for each unique value

ttn['embark_town'].value_counts()

Southampton    644
Cherbourg      168
Queenstown      77
Name: embark_town, dtype: int64

In [39]:
# Using the Groupby function
ttn.groupby('sex')

## Order of groupby changes the interpretation

In [84]:
# Survival based on different classes
summsurvived=ttn.groupby(['pclass','survived','sex'])['pclass'].count()
summsurvived

pclass  survived  sex   
1       0         female      3
                  male       77
        1         female     91
                  male       45
2       0         female      6
                  male       91
        1         female     70
                  male       17
3       0         female     72
                  male      300
        1         female     72
                  male       47
Name: pclass, dtype: int64

In [85]:
# Survival based on different sex
summsurvived=ttn.groupby(['sex','survived','pclass'])['pclass'].count()
summsurvived

sex     survived  pclass
female  0         1           3
                  2           6
                  3          72
        1         1          91
                  2          70
                  3          72
male    0         1          77
                  2          91
                  3         300
        1         1          45
                  2          17
                  3          47
Name: pclass, dtype: int64

a) More females in count survived ( females survived = 233 against male survived = 109). 
b) Percentage wise female survival was high (233/ 314 =74.2% against male (109/577 =18.89%)
c) Females who survived the class did not matter. Females who did not survive, most of them were from class 3
d) Males who survived the class did not matter. Males who did not survive, most of them were from class 3.


In [86]:
ttn[(ttn['sex'] =='female') & (ttn['age'] > 50) ]

survived  pclass     sex   age  sibsp  parch      fare embarked   class  \
11          1       1  female  58.0      0      0   26.5500        S   First   
15          1       2  female  55.0      0      0   16.0000        S  Second   
195         1       1  female  58.0      0      0  146.5208        C   First   
268         1       1  female  58.0      0      1  153.4625        S   First   
275         1       1  female  63.0      1      0   77.9583        S   First   
366         1       1  female  60.0      1      0   75.2500        C   First   
483         1       3  female  63.0      0      0    9.5875        S   Third   
496         1       1  female  54.0      1      0   78.2667        C   First   
513         1       1  female  54.0      1      0   59.4000        C   First   
571         1       1  female  53.0      2      0   51.4792        S   First   
591         1       1  female  52.0      1      0   78.2667        C   First   
765         1       1  female  51.0      1      0   77.9583        S   First   
772         0       2  female  57.0      0      0   10.5000        S  Second   
774         1       2  female  54.0      1      3   23.0000        S  Second   
820         1       1  female  52.0      1      1   93.5000        S   First   
829         1       1  female  62.0      0      0   80.0000      NaN   First   
879         1       1  female  56.0      0      1   83.1583        C   First   

       who  adult_male deck  embark_town alive  alone  
11   woman       False    C  Southampton   yes   True  
15   woman       False  NaN  Southampton   yes   True  
195  woman       False    B    Cherbourg   yes   True  
268  woman       False    C  Southampton   yes  False  
275  woman       False    D  Southampton   yes  False  
366  woman       False    D    Cherbourg   yes  False  
483  woman       False  NaN  Southampton   yes   True  
496  woman       False    D    Cherbourg   yes  False  
513  woman       False  NaN    Cherbourg   yes  False  
571  woman       False    C  Southampton   yes  False  
591  woman       False    D    Cherbourg   yes  False  
765  woman       False    D  Southampton   yes  False  
772  woman       False    E  Southampton    no   True  
774  woman       False  NaN  Southampton   yes  False  
820  woman       False    B  Southampton   yes  False  
829  woman       False    B          NaN   yes   True  
879  woman       False    C    Cherbourg   yes  False

In [91]:
# Groupby function with size gives the count
ttn.groupby(['sex','survived']).size()

sex     survived
female  0            81
        1           233
male    0           468
        1           109
dtype: int64

In [93]:
ttn.groupby(['sex','survived'])['pclass'].count()

sex     survived
female  0            81
        1           233
male    0           468
        1           109
Name: pclass, dtype: int64

In [94]:
# Median age of people who survived and did not survive
ttn.groupby('survived')['age'].median()

survived
0    28.0
1    28.0
Name: age, dtype: float64

In [96]:
# Data can be grouped by more than 1 column
ttn.groupby(['sex', 'survived']).count()

pclass  age  sibsp  parch  fare  embarked  class  who  \
sex    survived                                                          
female 0             81   64     81     81    81        81     81   81   
       1            233  197    233    233   233       231    233  233   
male   0            468  360    468    468   468       468    468  468   
       1            109   93    109    109   109       109    109  109   

                 adult_male  deck  embark_town  alive  alone  
sex    survived                                               
female 0                 81     6           81     81     81  
       1                233    91          231    233    233  
male   0                468    61          468    468    468  
       1                109    45          109    109    109

In [103]:
# Index 0 and the value of sex
ttn.loc[(0,'sex')]

'male'

In [104]:
# all the values in the row 0
ttn.loc[(0)]

survived                 0
pclass                   3
sex                   male
age                   22.0
sibsp                    1
parch                    0
fare                  7.25
embarked                 S
class                Third
who                    man
adult_male            True
deck                   NaN
embark_town    Southampton
alive                   no
alone                False
Name: 0, dtype: object

In [120]:
# Using series and finding out what is the sex through out the series. You can also do slicing by doing 'a':'b'
ttn.loc[1:5,'sex':'who']

sex   age  sibsp  parch     fare embarked  class    who
1  female  38.0      1      0  71.2833        C  First  woman
2  female  26.0      0      0   7.9250        S  Third  woman
3  female  35.0      1      0  53.1000        S  First  woman
4    male  35.0      0      0   8.0500        S  Third    man
5    male   NaN      0      0   8.4583        Q  Third    man